fluence-faas-service/src/service.rs

```diff
@@ -0,0 +1,139 @@
+
+use crate::Result;
+use super::ServiceError;
+use super::IValue;
+
+use fluence_faas::FluenceFaaS;
+use fluence_faas::ModulesConfig;
+
+use std::convert::TryInto;
+
+const SERVICE_ID_ENV_NAME: &str = "service_id";
+
+// TODO: remove and use mutex instead
+unsafe impl Send for FluenceFaaSService {}
+
+pub struct FluenceFaaSService {
+    faas: FluenceFaaS,
+}
+
+impl FluenceFaaSService {
+    /// Creates Service with given modules and service id.
+    pub fn new<I, C, S>(modules: I, config: C, service_id: S) -> Result<Self>
+    where
+        I: IntoIterator<Item = String>,
+        C: TryInto<ModulesConfig>,
+        S: AsRef<str>,
+        ServiceError: From<C::Error>,
+    {
+        let config: ModulesConfig = config.try_into()?;
+        let service_id = service_id.as_ref();
+        let config = Self::prepare_before_creation(config, service_id)?;
+
+        let modules = modules.into_iter().collect();
+        let faas = FluenceFaaS::with_module_names(&modules, config)?;
+
+        Ok(Self { faas })
+    }
+
+    /// Call a specified function of loaded module by its name.
+    // TODO: replace serde_json::Value with Vec<u8>?
+    pub fn call_module<MN: AsRef<str>, FN: AsRef<str>>(
+        &mut self,
+        module_name: MN,
+        func_name: FN,
+        arguments: serde_json::Value,
+    ) -> Result<Vec<IValue>> {
+        let arguments = Self::prepare_arguments(arguments)?;
+
+        self.faas
+            .call_module(module_name, func_name, &arguments)
+            .map_err(Into::into)
+    }
+
+    /// Return all export functions (name and signatures) of loaded modules.
+    pub fn get_interface(&self) -> fluence_faas::FaaSInterface<'_> {
+        self.faas.get_interface()
+    }
+
+    fn prepare_before_creation(
+        mut config: ModulesConfig,
+        service_id: &str,
+    ) -> Result<ModulesConfig> {
+        let base_dir = match config.service_base_dir {
+            Some(ref base_dir) => base_dir,
+            // TODO: refactor it later
+            None => {
+                return Err(ServiceError::IOError(String::from(
+                    "service_base_dir should be specified",
+                )))
+            }
+        };
+
+        let service_dir = std::path::Path::new(base_dir).join(service_id);
+        std::fs::create_dir(service_dir.clone())?; // will return an error if dir is already exists
+
+        let service_id_env = vec![format!("{}={}", SERVICE_ID_ENV_NAME, service_id).into_bytes()];
+        let preopened_files = vec![];
+        let mapped_dirs = vec![(
+            String::from("service_dir"),
+            service_dir.to_string_lossy().into(),
+        )];
+
+        config.modules_config = config
+            .modules_config
+            .into_iter()
+            .map(|(name, module_config)| {
+                let module_config = module_config
+                    .extend_wasi_envs(service_id_env.clone())
+                    .extend_wasi_files(preopened_files.clone(), mapped_dirs.clone());
```

Comment:
What about previously configured mapped_dirs? Can they leak out of service directory? e.g if in the config it's
```
[[modules]]
name = "A"
mapped_dirs = { "/tmp" -> "/tmp" }

name = "B"
mapped_dirs = { "/tmp" -> "/tmp" }
```
will A & B share the same "/tmp" directory?